lib/puppet/face/facts.rb

```diff
@@ -141,7 +141,7 @@
         end
         facter4_result = Puppet::Util::Execution.execute(puppet_cmd)
 
-        fact_diff = FactDif.new(facter3_result.to_json, facter4_result, EXCLUDE_LIST)
+        fact_diff = FactDif.new(facter3_result.values.to_json, facter4_result, EXCLUDE_LIST)
```

Comment:
Since you are passing the values, this should be updated not to fetch values from the resulted hash: https://github.com/ciprianbadescu/puppet/blob/maint/facts-diff-quickfix/lib/puppet/util/fact_dif.rb#L5-L6

Reply:
You might also need to pass `--show-legacy` in the puppet facts show command, as Facter 4 disables legacy facts by default.

Reply:
Also i think `puppet facts show` will be missing some facts, like `clientcert` due to https://github.com/puppetlabs/puppet/commit/c9dfabb6a395ca96c216b52d6686678e0b0751ee

Reply:
agree